src/rmind/components/llm.py

```diff
@@ -1,46 +1,116 @@
-)
-from xformers.components.reversible import ReversibleSequence
-from xformers.factory.model_factory import (
-    get_weight_init_fn,
-    xFormerEncoderBlock,
-    xFormerEncoderConfig,
-    xFormerWeightInit,
-)
-
-
-@final
-@register_feedforward("MLPGLU", MlpConfig)
-class MLPGLU(Feedforward):
+from torch.utils.checkpoint import checkpoint
+
+
+class TransformerEncoderBlock(nn.Module):
+    def __init__(  # noqa: PLR0913, PLR0917
+        self,
+        embedding_dim: int,
+        num_heads: int,
+        attn_dropout: float = 0.1,
+        resid_dropout: float = 0.1,
+        mlp_dropout: float = 0.1,
+        hidden_layer_multiplier: int = 1,
+    ) -> None:
+        super().__init__()
+
+        # f
+        self.pre_norm: nn.LayerNorm = nn.LayerNorm(embedding_dim)  # pre-norm
+
+        self.mha: nn.MultiheadAttention = nn.MultiheadAttention(
+            embed_dim=embedding_dim,
+            num_heads=num_heads,
+            dropout=attn_dropout,
+            batch_first=True,
+        )
+
+        self.resid_drop: nn.Dropout = nn.Dropout(resid_dropout, inplace=False)
+
+        # g
+        self.post_norm: nn.LayerNorm = nn.LayerNorm(embedding_dim)  # ffn
+
+        self.mlp: MLPGLU = MLPGLU(
+            dim_model=embedding_dim,
+            dropout=mlp_dropout,
+            activation="gelu",
+            hidden_layer_multiplier=hidden_layer_multiplier,
+        )
+
+    @override
+    def forward(self, x: Tensor, mask: Tensor) -> Tensor:
+        # f
+        residual = x
+        x_norm = self.pre_norm(x)
+        mha, _ = self.mha(x_norm, x_norm, x_norm, attn_mask=mask, need_weights=False)
+        x = residual + self.resid_drop(mha)
+
+        # g
+        residual = x
+        mlp = self.mlp(self.post_norm(x))
+        return residual + mlp
+
+
+class TransformerEncoder(nn.Module):
+    def __init__(  # noqa: PLR0913, PLR0917
+        self,
+        dim_model: int,
+        num_layers: int,
+        num_heads: int,
+        attn_dropout: float = 0.1,
+        resid_dropout: float = 0.1,
+        mlp_dropout: float = 0.1,
+        hidden_layer_multiplier: int = 1,
+    ) -> None:
+        super().__init__()
+        self.layers = nn.ModuleList([  # pyright: ignore[reportUnannotatedClassAttribute]
+            TransformerEncoderBlock(
+                embedding_dim=dim_model,
+                num_heads=num_heads,
+                attn_dropout=attn_dropout,
+                mlp_dropout=mlp_dropout,
+                resid_dropout=resid_dropout,
+                hidden_layer_multiplier=hidden_layer_multiplier,
+            )
+            for _ in range(num_layers)
+        ])
+        # https://github.com/karpathy/nanoGPT/blob/master/model.py#L182
+        self.layer_norm: nn.LayerNorm = nn.LayerNorm(dim_model)
+
+    @override
+    def forward(self, src: Tensor, mask: Tensor) -> Tensor:
+        x = src
+
+        for layer in self.layers:
+            if self.training:
```

Comment:
[nit] rather than checking `self.training` every time (presumably doesn't change as we iterate over layers), lift the if-guard up and check once?

Reply:
how so?

Reply:
you are right it doesn't